File: MyScript/edk/c/examples/shape/IncrementalShapeRecognitionBeautification.c
/**
 * The `IncrementalShapeRecognitionBeautification` example code shows you how to
 * perform incremental shape recognition and beautification.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptShape.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// digital ink
#include "rectangle.strokes.h"
#include "circle.strokes.h"
#include "line.strokes.h"
#include "arrow_head.strokes.h"

// forward declaration of example functions
static void printResult(voShapeDocument document);

// main function
int main()
{
  voShapeRecognizer recognizer = NULL;
  voShapeBeautifier beautifier = NULL;
  voShapeKnowledge shapeKnowledge = NULL;
  voShapeDocument document = NULL;

  // -- initialization ---------------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) creating a new ShapeRecognizer and a new ShapeBeautifier\n");
  recognizer = voCreateShapeRecognizer(engine);
  verify(recognizer != NULL);
  printf(" . ShapeRecognizer created successfully\n");
  beautifier = voCreateShapeBeautifier(engine);
  verify(beautifier != NULL);
  printf(" . ShapeBeautifier created successfully\n");
  printf("\n");

  printf("3) loading and attaching the ShapeKnowledge\n");
  shapeKnowledge = loadEngineObject(SHAPE_KNOWLEDGE_RES);
  printf(" . ShapeKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, shapeKnowledge));
  printf(" . ShapeKnowledge resource successfully attached to the ShapeRecognizer\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, beautifier, shapeKnowledge));
  printf(" . ShapeKnowledge resource successfully attached to the ShapeBeautifier\n");
  printf(" . releasing the ShapeKnowledge resource (not needed anymore)\n");
  verify(voReleaseObject(engine, shapeKnowledge));
  printf(" . ShapeKnowledge resource released successfully\n");
  printf("\n");

  printf("4) creating a ShapeDocument\n");
  document = voCreateShapeDocument(engine);
  verify(document != NULL);
  printf(" . ShapeDocument created successfully\n");
  printf("\n");

  printf("5) fueling the ShapeDocument with digital ink\n");
  printf(" . adding a rectangle\n");
  printf("   ,--------.\n");
  printf("   |        |\n");
  printf("   |        |\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s0[0], 2 * sizeof(*rectangle_s0), &rectangle_s0[1], 2 * sizeof(*rectangle_s0), (uint32_t)(sizeof(rectangle_s0) / sizeof(*rectangle_s0)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s1[0], 2 * sizeof(*rectangle_s1), &rectangle_s1[1], 2 * sizeof(*rectangle_s1), (uint32_t)(sizeof(rectangle_s1) / sizeof(*rectangle_s1)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s2[0], 2 * sizeof(*rectangle_s2), &rectangle_s2[1], 2 * sizeof(*rectangle_s2), (uint32_t)(sizeof(rectangle_s2) / sizeof(*rectangle_s2)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s3[0], 2 * sizeof(*rectangle_s3), &rectangle_s3[1], 2 * sizeof(*rectangle_s3), (uint32_t)(sizeof(rectangle_s3) / sizeof(*rectangle_s3)) / 2));

  printf(" . adding an ALMOST horizontal line\n");
  printf("   ,--------.    ,----\n");
  printf("   |        | ---\n");
  printf("   |        |\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &line_s0[0], 2 * sizeof(*line_s0), &line_s0[1], 2 * sizeof(*line_s0), (uint32_t)(sizeof(line_s0) / sizeof(*line_s0)) / 2));

  printf(" . adding a circle\n");
  printf("                         ,---.\n");
  printf("   ,--------.    ,----  /     \\\n");
  printf("   |        | ---      |       |\n");
  printf("   |        |           \\     /\n");
  printf("   '--------'            '---'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &circle_s0[0], 2 * sizeof(*circle_s0), &circle_s0[1], 2 * sizeof(*circle_s0), (uint32_t)(sizeof(circle_s0) / sizeof(*circle_s0)) / 2));
  printf(" . ShapeDocument successfully fueled with digital ink\n");
  printf("\n");

  printf("6) running the shape recognition and beautification processes\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, recognizer, document, NULL, NULL));
  printf(" . shape recognition performed successfully\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, beautifier, document, NULL, NULL));
  printf(" . shape beautification performed successfully\n");
  printf("\n");

  printf("7) retrieving the recognition and beautification result\n");
  printResult(document);
  printf("\n");

  printf("8) fueling the ShapeDocument with more digital ink\n");
  printf(" . adding an arrow head\n");
  printf(" . please note that in the typical incremental scenario, the user\n");
  printf("   draws the arrow head on top of a render of the previously recognized\n");
  printf("   and beautified diagram, the ink is not displayed anymore\n");
  printf(" . as you'll see next, the ink for the arrow head is in fact drawn\n");
  printf("   below the right extremity of that almost horizontal line\n");
  printf("                         ,---.\n");
  printf("   ,--------.           /     \\\n");
  printf("   |        | -------->|       |\n");
  printf("   |        |           \\     /\n");
  printf("   '--------'            '---'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &arrow_head_s0[0], 2 * sizeof(*arrow_head_s0), &arrow_head_s0[1], 2 * sizeof(*arrow_head_s0), (uint32_t)(sizeof(arrow_head_s0) / sizeof(*arrow_head_s0)) / 2));
  printf(" . ShapeDocument successfully fueled with digital ink\n");
  printf("\n");

  printf("9) running the shape recognition and beautification processes\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, recognizer, document, NULL, NULL));
  printf(" . shape recognition performed successfully\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, beautifier, document, NULL, NULL));
  printf(" . shape beautification performed successfully\n");
  printf("\n");

  printf("10) retrieving the recognition and beautification result\n");
  printResult(document);
  printf("\n");

  printf("11) clearing the ShapeDocument\n");
  verify(VO_IFACE(engine, IShapeDocument)->clear(engine, document));
  printf(" . ShapeDocument cleared successfully\n");
  printf("\n");

  printf("12) fueling the ShapeDocument with digital ink AS A WHOLE\n");
  printf(" . adding a rectangle\n");
  printf("   ,--------.\n");
  printf("   |        |\n");
  printf("   |        |\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s0[0], 2 * sizeof(*rectangle_s0), &rectangle_s0[1], 2 * sizeof(*rectangle_s0), (uint32_t)(sizeof(rectangle_s0) / sizeof(*rectangle_s0)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s1[0], 2 * sizeof(*rectangle_s1), &rectangle_s1[1], 2 * sizeof(*rectangle_s1), (uint32_t)(sizeof(rectangle_s1) / sizeof(*rectangle_s1)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s2[0], 2 * sizeof(*rectangle_s2), &rectangle_s2[1], 2 * sizeof(*rectangle_s2), (uint32_t)(sizeof(rectangle_s2) / sizeof(*rectangle_s2)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s3[0], 2 * sizeof(*rectangle_s3), &rectangle_s3[1], 2 * sizeof(*rectangle_s3), (uint32_t)(sizeof(rectangle_s3) / sizeof(*rectangle_s3)) / 2));

  printf(" . adding an ALMOST horizontal line\n");
  printf("   ,--------.    ,---- \n");
  printf("   |        | ---      \n");
  printf("   |        |\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &line_s0[0], 2 * sizeof(*line_s0), &line_s0[1], 2 * sizeof(*line_s0), (uint32_t)(sizeof(line_s0) / sizeof(*line_s0)) / 2));

  printf(" . adding a circle\n");
  printf("                         ,---.\n");
  printf("   ,--------.    ,----  /     \\\n");
  printf("   |        | ---      |       |\n");
  printf("   |        |           \\     /\n");
  printf("   '--------'            '---'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &circle_s0[0], 2 * sizeof(*circle_s0), &circle_s0[1], 2 * sizeof(*circle_s0), (uint32_t)(sizeof(circle_s0) / sizeof(*circle_s0)) / 2));

  printf(" . adding an arrow head (as if the line was horizontal)\n");
  printf("                         ,---.\n");
  printf("   ,--------.    ,----  /     \\\n");
  printf("   |        | ---     >|       |\n");
  printf("   |        |           \\     /\n");
  printf("   '--------'            '---'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &arrow_head_s0[0], 2 * sizeof(*arrow_head_s0), &arrow_head_s0[1], 2 * sizeof(*arrow_head_s0), (uint32_t)(sizeof(arrow_head_s0) / sizeof(*arrow_head_s0)) / 2));
  printf(" . ShapeDocument successfully fueled with digital ink\n");
  printf("\n");

  printf("13) running the shape recognition and beautification processes\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, recognizer, document, NULL, NULL));
  printf(" . shape recognition performed successfully\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, beautifier, document, NULL, NULL));
  printf(" . shape beautification performed successfully\n");
  printf("\n");

  printf("14) retrieving the recognition result\n");
  printResult(document);
  printf("\n");

  printf("15) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, document));
  verify(voReleaseObject(engine, recognizer));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform incremental recognition and beautification\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}

static void printResult(voShapeDocument document)
{
  uint32_t segmentCount = 0;
  uint32_t i = 0;

  printf(" . retrieving the number of ShapeSegments in the ShapeDocument: ");
  segmentCount = VO_IFACE(engine, IShapeDocument)->getSegmentCount(engine, document);
  verify(segmentCount != (uint32_t)-1);
  printf("%d\n", segmentCount);

  printf(" . iterating over the ShapeSegments:\n");
  for (i = 0; i < segmentCount; ++i)
  {
    voShapeSegment segment = NULL;
    uint32_t candidateCount = 0;

    segment = VO_IFACE(engine, IShapeDocument)->getSegmentAt(engine, document, i);
    verify(segment != NULL);

    candidateCount = VO_IFACE(engine, IShapeSegment)->getCandidateCount(engine, segment);
    verify(candidateCount != (uint32_t)-1);

    if (candidateCount != 0)
    {
      voShapeCandidate candidate = NULL;

      // get the best candidate
      candidate = VO_IFACE(engine, IShapeSegment)->getCandidateAt(engine, segment, 0);
      verify(candidate != NULL);

      switch(voGetType(engine, candidate))
      {
        case VO_ShapeRecognized:
          {
            voShapeModel model = NULL;
            voString label = {NULL, 0};

            float rs = 0;
            float nrs = 0;

            model = VO_IFACE(engine, IShapeRecognized)->getModel(engine, candidate);
            verify(model != NULL);

            verify(VO_IFACE(engine, IShapeModel)->getLabel(engine, model, NULL, &label));
            label.bytes = malloc(label.byteCount + 1);
            verify(VO_IFACE(engine, IShapeModel)->getLabel(engine, model, NULL, &label));
            label.bytes[label.byteCount] = '\0';

            rs = VO_IFACE(engine, IShapeRecognized)->getResemblanceScore(engine, candidate);
            nrs = VO_IFACE(engine, IShapeRecognized)->getNormalizedRecognitionScore(engine, candidate);

            printf("   . segment %d, best candidate: %s (rs:%.4f - nrs:%.4f)\n", i, label.bytes, rs, nrs);

            free(label.bytes);
            // don't forget to release objects returned by the API once you're done
            verify(voReleaseObject(engine, model));
          }
          break;

        case -1:
          printf("unable to get candidate type\n");
          break;

        default:
          // ignore other types of candidates
          break;
      }

      verify(voReleaseObject(engine, candidate));
    }

    verify(voReleaseObject(engine, segment));
  }
}

File: MyScript/edk/java/examples/shape/IncrementalShapeRecognitionBeautification.java
package shape;

import java.text.NumberFormat;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.shape.ShapeBeautifier;
import com.myscript.shape.ShapeCandidate;
import com.myscript.shape.ShapeDocument;
import com.myscript.shape.ShapeKnowledge;
import com.myscript.shape.ShapeModel;
import com.myscript.shape.ShapeRecognized;
import com.myscript.shape.ShapeRecognizer;
import com.myscript.shape.ShapeSegment;
import common.CFG;

/**
 * The <code>IncrementalShapeRecognitionBeautification</code> example code shows
 * you how to perform incremental shape recognition and beautification.
 */
public class IncrementalShapeRecognitionBeautification
{
  /**
   * Main entry point for the
   * <code>IncrementalShapeRecognitionBeautification</code> example program.
   *
   * @param args command line arguments, ignored.
   */
  public static final void main(String[] args)
  {
    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) creating a new ShapeRecognizer and a new ShapeBeautifier");
    final ShapeRecognizer recognizer = ShapeRecognizer.create(engine);
    System.out.println(" . ShapeRecognizer created successfully");
    final ShapeBeautifier beautifier = ShapeBeautifier.create(engine);
    System.out.println(" . ShapeBeautifier created successfully");
    System.out.println();

    System.out.println("3) loading and attaching the ShapeKnowledge to the ShapeRecognizer");
    final ShapeKnowledge shapeKnowledge = (ShapeKnowledge)EngineObject.load(engine, CFG.getShapeKnowledgeRes());
    System.out.println(" . ShapeKnowledge resource loaded successfully");
    recognizer.attach(shapeKnowledge);
    System.out.println(" . ShapeKnowledge resource attached successfully to the ShapeRecognizer");
    beautifier.attach(shapeKnowledge);
    System.out.println(" . ShapeKnowledge resource attached successfully to the ShapeBeautifier");
    System.out.println(" . disposing of the ShapeKnowledge resource (not needed anymore)");
    shapeKnowledge.dispose();
    System.out.println(" . ShapeKnowledge resource disposed of successfully");
    System.out.println();
    
    System.out.println("4) creating a new ShapeDocument");
    final ShapeDocument document = ShapeDocument.create(engine);
    System.out.println(" . ShapeDocument created successfully");
    System.out.println();

    System.out.println("5) fueling the ShapeDocument with digital ink");
    System.out.println(" . adding a rectangle");
    System.out.println("   ,--------.");
    System.out.println("   |        |");
    System.out.println("   |        |");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.RECTANGLE.length; ++i)
    {
      float[] xy = Ink.RECTANGLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding an ALMOST horizontal line");
    System.out.println("   ,--------.    ,----");
    System.out.println("   |        | ---");
    System.out.println("   |        |");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.LINE.length; ++i)
    {
      float[] xy = Ink.LINE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding a circle");
    System.out.println("                         ,---.");
    System.out.println("   ,--------.    ,----  /     \\");
    System.out.println("   |        | ---      |       |");
    System.out.println("   |        |           \\     /");
    System.out.println("   '--------'            '---'");
    for (int i = 0; i < Ink.CIRCLE.length; ++i)
    {
      float[] xy = Ink.CIRCLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    System.out.println(" . ShapeDocument successfully fueled with digital ink");
    System.out.println();
    
    System.out.println("6) running the shape recognition and beautification process");
    recognizer.process(document);
    System.out.println(" . shape recognition performed successfully");
    beautifier.process(document);
    System.out.println(" . shape beautification performed successfully");
    System.out.println();
    
    System.out.println("7) retrieving the recognition and beautification result");
    printResult(document);
    System.out.println();

    System.out.println("8) fueling the ShapeDocument with more digital ink");
    System.out.println(" . adding an arrow head");
    System.out.println(" . please note that in the typical incremental scenario, the user");
    System.out.println("   draws the arrow head on top of a render of the previously recognized");
    System.out.println("   and beautified diagram, the ink is not displayed anymore");
    System.out.println(" . as you'll see next, the ink for the arrow head is in fact drawn");
    System.out.println("   below the right extremity of that almost horizontal line");
    System.out.println("                         ,---.");
    System.out.println("   ,--------.           /     \\");
    System.out.println("   |        | -------->|       |");
    System.out.println("   |        |           \\     /");
    System.out.println("   '--------'            '---'");
    for (int i = 0; i < Ink.ARROW_HEAD.length; ++i)
    {
      float[] xy = Ink.ARROW_HEAD[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    System.out.println(" . ShapeDocument successfully fueled with digital ink");
    System.out.println();
    
    System.out.println("9) running the shape recognition and beautification process");
    recognizer.process(document);
    System.out.println(" . shape recognition performed successfully");
    beautifier.process(document);
    System.out.println(" . shape beautification performed successfully");
    System.out.println();
    
    System.out.println("10) retrieving the recognition and beautification result");
    printResult(document);
    System.out.println();

    System.out.println("11) clearing the ShapeDocument");
    document.clear();
    System.out.println(" . ShapeDocument cleared successfully");
    System.out.println();
    
    System.out.println("12) fueling the ShapeDocument with digital ink AS A WHOLE");
    System.out.println(" . adding a rectangle");
    System.out.println("   ,--------.");
    System.out.println("   |        |");
    System.out.println("   |        |");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.RECTANGLE.length; ++i)
    {
      float[] xy = Ink.RECTANGLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding an ALMOST horizontal line");
    System.out.println("   ,--------.    ,----");
    System.out.println("   |        | ---");
    System.out.println("   |        |");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.LINE.length; ++i)
    {
      float[] xy = Ink.LINE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding a circle");
    System.out.println("                         ,---.");
    System.out.println("   ,--------.    ,----  /     \\");
    System.out.println("   |        | ---      |       |");
    System.out.println("   |        |           \\     /");
    System.out.println("   '--------'            '---'");
    for (int i = 0; i < Ink.CIRCLE.length; ++i)
    {
      float[] xy = Ink.CIRCLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding an arrow head (as if the line was horizontal)");
    System.out.println("                         ,---.");
    System.out.println("   ,--------.    ,----  /     \\");
    System.out.println("   |        | ---     >|       |");
    System.out.println("   |        |           \\     /");
    System.out.println("   '--------'            '---'");
    for (int i = 0; i < Ink.ARROW_HEAD.length; ++i)
    {
      float[] xy = Ink.ARROW_HEAD[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    System.out.println(" . ShapeDocument successfully fueled with digital ink");
    System.out.println();
    
    System.out.println("13) running the shape recognition and beautification process");
    recognizer.process(document);
    System.out.println(" . shape recognition performed successfully");
    beautifier.process(document);
    System.out.println(" . shape beautification performed successfully");
    System.out.println();
    
    System.out.println("14) retrieving the recognition and beautification result");
    printResult(document);
    System.out.println();
    
    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("15) cleaning-up: disposing of objects and disposing of the Engine");
    document.dispose();
    recognizer.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();
    
    System.out.println("end of example:");
    System.out.println(" you just learnt how to perform incremental recognition and beautification");
  }
  
  private static void printResult(final ShapeDocument document)
  {
    System.out.print(" . retrieving the number of ShapeSegments in the ShapeDocument: ");
    final int segmentCount = document.getSegmentCount();
    System.out.println(segmentCount);
    
    final NumberFormat nf = NumberFormat.getInstance();
    nf.setGroupingUsed(false);
    
    System.out.println(" . iterating over the ShapeSegments:");
    for (int i = 0; i < segmentCount; ++i)
    {
      final ShapeSegment segment = document.getSegmentAt(i);
      final int candidateCount = segment.getCandidateCount();

      if (candidateCount != 0)
      {
        final ShapeCandidate candidate = segment.getCandidateAt(0);

        if (candidate instanceof ShapeRecognized)
        {
          final ShapeModel model = ((ShapeRecognized)candidate).getModel();

          final float rs = ((ShapeRecognized)candidate).getResemblanceScore();
          final float nrs = ((ShapeRecognized)candidate).getNormalizedRecognitionScore();

          nf.setMinimumFractionDigits(4);
          nf.setMaximumFractionDigits(4);
          System.out.println("   . segment " + i + ", best candidate: " + model.getLabel() +" (rs:" + nf.format(rs) + " nrs:" + nf.format(nrs) + ')');
          
          model.dispose();
        }

        candidate.dispose();
      }

      segment.dispose();
    }
  }
  
} // MultipleShapesRecognitionBeautification
